R/credentials.R

```diff
@@ -0,0 +1,365 @@
+#' # secure key-value store
+#' create_smtp_creds_key(
+#'   provider = "gmail",
+#'   user = "user_name@gmail.com",
+#'   )
+#' }
+#' @export
+create_smtp_creds_key <- function(key_name = NULL,
+                                  provider = NULL,
+                                  user,
+                                  sender_name = NULL,
+                                  host = NULL,
+                                  port = NULL,
+                                  use_ssl = TRUE,
+                                  ...,
+                                  password = NULL) {
+
+  if (!keyring::has_keyring_support()) {
+    stop("To store SMTP via *keyring*, the system needs to have",
+         "*keyring* support", call. = FALSE)
+  }
+
+  # Ensure that `use_ssl` is either TRUE or FALSE
+  if (!(use_ssl %in% c(TRUE, FALSE))) {
+    stop("The value supplied to `use_ssl` must be TRUE or FALSE.")
+  }
+
+  if (!is.null(provider) &&
+      provider %in% (smtp_settings() %>% dplyr::pull(short_name))) {
+
+    # Extract the record for the SMTP provider
+    settings_record <-
+      smtp_settings() %>%
+      dplyr::filter(short_name == provider)
+
+    # TODO: only set these if the incoming values are not NULL
+    # Extract settings for the provider
+    host <- settings_record$server
+    port <- settings_record$port
+    use_ssl <- settings_record$use_ssl
+  }
+
+  # If the `sender_name` isn't provided, use
+  # an empty string
+  if (is.null(sender_name)) sender_name <- ""
+
+  # Use the provider name if it's given as the
+  # `key_name` (if that's not given)
+  if (is.null(key_name) && !is.null(provider)) {
+    key_name <- provider
+  } else if (is.null(key_name)) {
+    key_name <- ""
+  }
+
+  # Create the `service_name` for `keyring`
+  service_name <- paste0("blastula-v", keyring_schema_version, "-", key_name)
+
+  credentials_list <-
+    list(
+      version = keyring_schema_version,
+      sender_name = sender_name,
+      host = host,
+      port = port,
+      use_ssl = use_ssl,
+      user = user,
+      password = password %||% getPass::getPass("Enter the SMTP server password: ")
+    )
+
+  serialized <-
+    credentials_list %>%
+    jsonlite::serializeJSON() %>%
+    as.character()
+
+  # Set the key in the system's default keyring
+  keyring::key_set_with_value(
+    service = service_name,
+    user = user,
+    password = serialized
+  )
+
+  # Issue a message stating that the file has been created
+  message("The system key store has been updated with the (`", service_name, "`) key.")
+}
+
+#' Utility function for obtaining keyring entries related to blastula creds
+#'
+#' @import keyring
+#' @importFrom dplyr as_tibble filter mutate
+#' @importFrom tidyr separate
+#' @noRd
+get_keyring_creds_table <- function(version = keyring_schema_version) {
+
+  creds_tbl <-
+    keyring::key_list() %>%
+    dplyr::as_tibble() %>%
+    dplyr::filter(grepl(paste0("blastula-v", version), service))
```

Comment:
Can you just `key_get(paste0("blastula-v", version), service)`? And also are you not missing a dash before `service`?

Reply:
Never mind, I get it